cucumber/index.d.ts

```diff
@@ -70,5 +70,139 @@ declare namespace cucumber {
 		Around(code: AroundCode):void;
 		setDefaultTimeout(time:number): void;
 		registerHandler(handlerOption:string, code:(event:any, callback:CallbackStepDefinition) =>void): void;
+		registerListener(listener: EventListener): void;
+	}
+
+	export class EventListener {
+		hear(event: events.Event, callback: ()=>void): void;
+		hasHandlerForEvent(event: events.Event): boolean;
+		buildHandlerNameForEvent(event: events.Event): string;
+		getHandlerForEvent(event: events.Event): EventHook;
+		buildHandlerName(shortName: string): string;
+		setHandlerForEvent(shortName: string, handler: EventListener): void;
+	}
+
+	export function Listener(): EventListener;
+
+	export namespace events {
+
+		interface Event {
+			getName(): string;
+			getPayloadItem(name: string): EventPayload;
+		}
+
+		interface EventPayload {
```

Comment:
Well, yes and no :-) It seems like a bit more context is needed here.

To notify a `Listener` of different phases in the test scenario lifecycle, Cucumber emits events which contain a field of "type" `EventPayload`, which value you can retrieve via the `getEventPayload(name)` method on the `Event` object. Regretfully, those different types of payloads have nothing in common API-wise, only the fact that they're passed as part of the event. 

I used the marker interface to:
- satisfy the signature to `Event::getPayloadItem(name: string): EventPayload` 
- highlight that the method can return all those different kinds of events

An alternative approach would be to define the method signature as returning `any`, remove the marker interface and disassociate the payloads from each other since they don't have anything in common design-wise, but I didn't find it convincing. I could be missing other alternatives though so I'm open to suggestions!

The nice thing about having both the method signature and the payloads typed though is that you can define a listener like this:

```typescript

let listener = Object.assign(cucumber.Listener(), {
    handleBeforeScenarioEvent,
    // ...
});

function handleBeforeScenarioEvent (scenario: cucumber.events.ScenarioPayload, callback: () => void) {
    
    // stuff

    callback();
}
```